### DISCUSSION TITLE
PA2: about format of packet

### DISCUSSION THREAD

**Discussion by overflow218** [ANLAB-KAIST/KENSv3#59] [Project Q&A 2. Transport Layer (TCP)] [answered]
Dear TA, I'm trying to implement connect. 
I read the tip: sending and receiving packet and tried to use it. 
I made a TCP header packet with format with Text book.
but when I use sendPacket() function with this packet, It stuck on assertion    assert(packet.getSize() >= ip_start + 20);
should consider Ip layer and make a room for Ip datagram at TCP header packet? 
-----------------------
안녕하세요 조교님 
connect 구현을 위해 교과서를 참고하여 TCP header packet을 만들었습니다. 
그런데 sendPacket()에 패킷을 담아 보내려고하면  
 assert(packet.getSize() >= ip_start + 20); 
와 같은 에러가 발생합니다. 패킷의 사이즈가 최소 ip_start + 20 은 되어야한다는 의미인것 같은데, ip layer 부분도 고려해야하는 건가요?
그러면 패킷 데이터의 0 ~ 13 부분에 들어가는 내용도 고려해주어야하는 것인가요? 아니면 그냥 
크기만 크게 해놓고 데이터는 앞에서부터 TCP header packet 의 내용을 채워주면 되는 것인가요?
감사합니다.

**Comment by Pusnow** [comment, marked as answer]
Yes, you should reserve space for the Ethernet and IP headers.
But you don't fill all of their fields.
You only need to specify source and destination IP addresses.

The other fields are filled in

https://github.com/ANLAB-KAIST/KENSv3/blob/c95d8f9a7669a491317d53eb4ff2db96ad8a0ffc/src/Networking/IPv4/E_IPv4.cpp#L67-L123
